File: rb-appscript/tags/rb-appscript-0.1.5/test/test_aemreference.rb
#!/usr/local/bin/ruby

require 'test/unit'
require "_aem/aemreference"
require "_aem/typewrappers"
require "_aem/codecs"


class TC_AEMReferences < Test::Unit::TestCase

	def test_referenceForms
		[
			[AEMReference::App.property('ctxt'), 'AEM.app.property("ctxt")', nil],
			
			[AEMReference::App.elements('docu'), 'AEM.app.elements("docu")', nil],
			
			[AEMReference::App.elements('docu').byindex(1), 
					'AEM.app.elements("docu").byindex(1)', nil],
			[AEMReference::App.elements('docu').byname('foo'), 
					'AEM.app.elements("docu").byname("foo")', nil],
			[AEMReference::App.elements('docu').byid(300), 
					'AEM.app.elements("docu").byid(300)', nil],
			
			[AEMReference::App.elements('docu').next('docu'), 
					'AEM.app.elements("docu").next("docu")', nil],
			[AEMReference::App.elements('docu').previous('docu'), 
					'AEM.app.elements("docu").previous("docu")', nil],
			
			[AEMReference::App.elements('docu').first, 'AEM.app.elements("docu").first', nil],
			[AEMReference::App.elements('docu').middle, 'AEM.app.elements("docu").middle', nil],
			[AEMReference::App.elements('docu').last, 'AEM.app.elements("docu").last', nil],
			[AEMReference::App.elements('docu').any, 'AEM.app.elements("docu").any', nil],
			
			[AEMReference::Con.elements("docu").byindex(3), 'AEM.con.elements("docu").byindex(3)', nil],
			
			[AEMReference::App.elements('docu').byrange(
					AEMReference::Con.elements('docu').byindex(3),
					AEMReference::Con.elements('docu').byname('foo')), 
					'AEM.app.elements("docu").byrange(' +
					'AEM.con.elements("docu").byindex(3), ' +
					'AEM.con.elements("docu").byname("foo"))', nil],
			
			
			[AEMReference::Its.property('name').eq('foo').and(AEMReference::Its.elements('cwor').eq([])), 
					'AEM.its.property("name").eq("foo").and(AEM.its.elements("cwor").eq([]))', nil],
			
			[AEMReference::Its.elements('cwor').ne([]), 
					'AEM.its.elements("cwor").ne([])', 
					AEMReference::Its.elements("cwor").eq([]).not], # i.e. there isn't a KAENotEqual operator, so not-equal tests are actually packed as an equal test followed by not test
					
			[AEMReference::Its.elements('cwor').eq(nil), 'AEM.its.elements("cwor").eq(nil)', nil],
			[AEMReference::Its.elements('cwor').property('leng').gt(0), 
					'AEM.its.elements("cwor").property("leng").gt(0)', nil],
			[AEMReference::Its.elements('cwor').le(''), 'AEM.its.elements("cwor").le("")', nil],
			[AEMReference::Its.elements('cwor').startswith('foo').not, 
					'AEM.its.elements("cwor").startswith("foo").not', nil],
			
			
			[AEMReference::Its.elements('cwor').contains('foo'), 'AEM.its.elements("cwor").contains("foo")', nil],
			[AEMReference::Its.elements('cwor').isin('foo'), 'AEM.its.elements("cwor").isin("foo")', nil],
			
			[AEMReference::App.elements('docu').byfilter(AEMReference::Its.property('size').ge(42)), 
					'AEM.app.elements("docu").byfilter(AEM.its.property("size").ge(42))', nil],
					
			[AEMReference::App.elements('docu').byindex(1).property('ctxt') \
					.elements('cpar').elements('cha ').byrange(
							AEMReference::Con.elements('cha ').byindex(3), 
							AEMReference::Con.elements('cha ').byindex(55)
					).next('cha ').after,
					'AEM.app.elements("docu").byindex(1).property("ctxt").elements("cpar").elements("cha ")' +
					'.byrange(AEM.con.elements("cha ").byindex(3), AEM.con.elements("cha ").byindex(55))' +
					'.next("cha ").after', nil],
			
			[AEMReference::Its.property('pnam').ne('foo').and(AEMReference::Its.elements('cfol').eq([])).not,
					'AEM.its.property("pnam").ne("foo").and(AEM.its.elements("cfol").eq([])).not',
					AEMReference::Its.property('pnam').eq('foo').not.and(AEMReference::Its.elements('cfol').eq([])).not],
			
			[AEMReference::App.elements('docu').start, 'AEM.app.elements("docu").start', nil],
			[AEMReference::App.elements('docu').end, 'AEM.app.elements("docu").end', nil],
			[AEMReference::App.elements('docu').byindex(3).before, 'AEM.app.elements("docu").byindex(3).before', nil],
			[AEMReference::App.elements('docu').byname('foo').after, 'AEM.app.elements("docu").byname("foo").after', nil],
			
		].each do |val, res, unpackedVersion|
			assert_equal(res, val.to_s)
			d = DefaultCodecs.pack(val)
			val = unpackedVersion ? unpackedVersion : val
			val2 = DefaultCodecs.unpack(d)
			assert_equal(val, val2)
			assert_block { val.eql?(val2) }
		end
		# by-range and by-filter references do basic type checking to ensure a reference is given
		assert_raises(TypeError) { AEMReference::App.elements('docu').byrange(1, 2) }
		assert_raises(TypeError) { AEMReference::App.elements('docu').byfilter(1) }
		
	end
end

File: py-appscript/tags/py-appscript-0.21.1/appscript_3x/test/test_aemreference.py
#!/usr/bin/env python3

import unittest
import aem

DefaultCodecs = aem.Codecs()


class TC_AEMReferences(unittest.TestCase):

	def test_referenceForms(self):
		for val, res, unpackedVersion in [
				[aem.app.property(b"ctxt"), "app.property(b'ctxt')", None],
				
				[aem.app.elements(b"docu"), "app.elements(b'docu')", None],
				
				[aem.app.elements(b"docu").byindex(1), 
						"app.elements(b'docu').byindex(1)", None],
				[aem.app.elements(b"docu").byname("foo"), 
						"app.elements(b'docu').byname('foo')", None],
				[aem.app.elements(b"docu").byid(300), 
						"app.elements(b'docu').byid(300)", None],
				
				[aem.app.elements(b"docu").next(b"docu"), 
						"app.elements(b'docu').next(b'docu')", None],
				[aem.app.elements(b"docu").previous(b"docu"), 
						"app.elements(b'docu').previous(b'docu')", None],
				
				[aem.app.elements(b"docu").first, "app.elements(b'docu').first", None],
				[aem.app.elements(b"docu").middle, "app.elements(b'docu').middle", None],
				[aem.app.elements(b"docu").last, "app.elements(b'docu').last", None],
				[aem.app.elements(b"docu").any, "app.elements(b'docu').any", None],
				
				[aem.con.elements(b'docu').byindex(3), "con.elements(b'docu').byindex(3)", None],
				
				[aem.app.elements(b"docu").byrange(
						aem.con.elements(b"docu").byindex(3),
						aem.con.elements(b"docu").byname("foo")), 
						"app.elements(b'docu').byrange(" +
						"con.elements(b'docu').byindex(3), " +
						"con.elements(b'docu').byname('foo'))", None],
								
				[aem.its.property(b"name").eq("foo").AND(aem.its.elements(b"cwor").eq([])), 
						"its.property(b'name').eq('foo').AND(its.elements(b'cwor').eq([]))", None],
				
				[aem.its.elements(b"cwor").ne([]), 
						"its.elements(b'cwor').ne([])", 
						aem.its.elements(b'cwor').eq([]).NOT], # i.e. there isn"t a kAENotEqual operator, so not-equal tests are actually packed as an equal test followed by not test
						
				[aem.its.elements(b"cwor").eq(None), "its.elements(b'cwor').eq(None)", None],
				[aem.its.elements(b"cwor").property(b"leng").gt(0), 
						"its.elements(b'cwor').property(b'leng').gt(0)", None],
				[aem.its.elements(b"cwor").le(""), "its.elements(b'cwor').le('')", None],
				[aem.its.elements(b"cwor").beginswith("foo").NOT, 
						"its.elements(b'cwor').beginswith('foo').NOT", None],
				
				
				[aem.its.elements(b"cwor").contains("foo"), "its.elements(b'cwor').contains('foo')", None],
				[aem.its.elements(b"cwor").isin("foo"), "its.elements(b'cwor').isin('foo')", None],
				
				[aem.app.elements(b"docu").byfilter(aem.its.property(b"size").ge(42)), 
						"app.elements(b'docu').byfilter(its.property(b'size').ge(42))", None],
						
				[aem.app.elements(b"docu").byindex(1).property(b"ctxt") \
						.elements(b"cpar").elements(b"cha ").byrange(
								aem.con.elements(b"cha ").byindex(3), 
								aem.con.elements(b"cha ").byindex(55)
						).next(b"cha ").after,
						"app.elements(b'docu').byindex(1).property(b'ctxt').elements(b'cpar').elements(b'cha ')" +
						".byrange(con.elements(b'cha ').byindex(3), con.elements(b'cha ').byindex(55))" +
						".next(b'cha ').after", None],
				
				[aem.its.property(b"pnam").ne("foo").AND(aem.its.elements(b"cfol").eq([])).NOT,
						"its.property(b'pnam').ne('foo').AND(its.elements(b'cfol').eq([])).NOT",
						aem.its.property(b"pnam").eq("foo").NOT.AND(aem.its.elements(b"cfol").eq([])).NOT],
				
				[aem.app.elements(b"docu").beginning, "app.elements(b'docu').beginning", None],
				[aem.app.elements(b"docu").end, "app.elements(b'docu').end", None],
				[aem.app.elements(b"docu").byindex(3).before, "app.elements(b'docu').byindex(3).before", None],
				[aem.app.elements(b"docu").byname("foo").after, "app.elements(b'docu').byname('foo').after", None],
				
				]:
			self.assertEqual(res, str(val))
			d = DefaultCodecs.pack(val)
			val = unpackedVersion and unpackedVersion or val
			#print val, d
			val2 = DefaultCodecs.unpack(d)
			self.assertEqual(val, val2)
			val2 = DefaultCodecs.unpack(d)
			self.assertEqual(val2, val)
		self.assertNotEqual(aem.app.property(b'ctxt').property(b'ctxt'), aem.con.property(b'ctxt').property(b'ctxt'))
		self.assertNotEqual(aem.app.property(b'foob').property(b'ctxt'), aem.app.property(b'ctxt').property(b'ctxt'))
		self.assertNotEqual(aem.app.elements(b'ctxt').property(b'ctxt'), aem.app.property(b'ctxt').property(b'ctxt'))
		self.assertNotEqual(aem.app.elements(b'ctxt').property(b'ctxt'), 333)
		self.assertNotEqual(333, aem.app.property(b'ctxt').property(b'ctxt'))
		# by-filter references do basic type checking to ensure a reference is given
		self.assertRaises(TypeError, aem.app.elements(b'docu').byfilter, 1)




if __name__ == '__main__':
	unittest.main()
